design-documents/frontend/admin-guidelines.md

```diff
@@ -0,0 +1,215 @@
+  - [i18n](#i18n)
+  - [Theming](#theming)
+  - [GraphQL client](#graphql-client)
+  - [TypeScript import helpers](#typescript-import-helpers)
+- [devDependencies](#devdependencies)
+  - [Craco](#craco)
+  - [Redux mock store](#redux-mock-store)
+  - [Testing library](#testing-library)
+
+### version
+
+The version of this `package.json` follows [semver](https://semver.org/) and will change with this document accordingly.
+
+```json
+"version": "0.1.0",
+```
+
+### scripts
+
+Customize your CRA configuration with [craco][].
+
+- [Don't eject](https://create-react-app.dev/docs/alternatives-to-ejecting).
+- Avoid forking, because you can oftentimes do the same thing by creating your own [craco][] plugin. See [craco plugins on npm](https://www.npmjs.com/search?q=craco%20plugin) and the [craco recipes](https://github.com/sharegate/craco/tree/master/recipes) for guidance.
+
+```json
+"scripts": {
+  "start": "craco start",
+  "build": "craco build",
+  "test": "craco test"
+},
+```
+
+### [browserslist](https://github.com/browserslist/browserslist#readme)
+
+Don't forget your [polyfills](https://create-react-app.dev/docs/supported-browsers-features#supported-language-features)!
+
+```json
+"browserslist": [
+  ">0.2%",
+  "not dead",
+  "not ie < 11",
+  "not op_mini all"
+],
+```
+
+Refer also to [Magento DevDocs | Supported browsers](https://devdocs.magento.com/guides/v2.3/install-gde/system-requirements_browsers.html).
+
+### dependencies
+
+#### Utility functions
+
+Use vanilla (native) JS if possible, utility functions if not. [You don't (may not) need Lodash/Underscore](https://github.com/you-dont-need/You-Dont-Need-Lodash-Underscore#you-dont-may-not-need-lodashunderscore-). See also [Just](https://github.com/angus-c/just#just) and [Queso](https://github.com/jedmao/queso#readme) for 1st-class TypeScript support.
+
+To be clear, if you adhere to the advice above and you still need some utility functions, it should only be a handful of them. Use your best judgment. Install what you need, but keep it slim!
+
+```json
+"@queso/camel-case": "^0",
+"@queso/kebab-case": "^1",
+```
+
+:point_up: just an example.
+
+#### SSR
+
+Prepare for [SSR](https://medium.com/walmartlabs/the-benefits-of-server-side-rendering-over-client-side-rendering-5d07ff2cefe8) by using [universal fetching](https://www.npmjs.com/package/ky-universal). Don't use [axios](https://www.npmjs.com/package/axios) or [fetch](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API) directly.
+
+```json
+"ky": "^0",
+"ky-universal": "^0",
+```
+
+#### React
+
+```json
+"react": "^16",
+"react-dom": "^16",
+// pending https://github.com/nfl/react-helmet/issues/465
+"react-helmet-async": "^1",
+```
+
+#### Routing
+
+See [The Future of React Router and @reach/router](https://reacttraining.com/blog/reach-react-router-future/). Basically, the new API will be much closer to `@reach/router` than React Router, so this prepares us for that future.
+
+```json
+"@reach/router": "^1",
```

Comment:
Have you found a way you like to make @reach/router support hash routing?

Reply:
No. Is that an important feature for you?

Reply:
I think so. We'll need client-side routing within the React app that doesn't interfere with Magento's built-in server-side routing. I would use a hash router for that. Did you have a different approach in mind?

Reply:
Short answer is no, has routing is not supported and seeing as react-router is going to merge with reach router, I guess we'll have to see if it survives the merge. Take a look at [this thread](https://github.com/reach/router/issues/25) that suggests using a memory router; though, also points out the fact that it won't survive a page refresh.

Reply:
There are two big reasons to use reach router:
1. The API is most compatible with the future merge of react router.
1. Nested routes are way nicer.

That said, if you really can't make it work with reach router, I guess you'll have to use react router (for now), understanding that the API will change significantly and you might even lose your hash routing capabilities, preventing you from upgrading it.

No great answer here. It's a double edged ⚔️

Reply:
Yeah I'd read that thread when I was trying to figure out how to solve this issue. I think the value of having the hash router now is worth a slightly more involved migration later on to my team.